homeassistant/components/media_player/template_media_player.py

```diff
@@ -0,0 +1,197 @@
+            CONF_VOLUME_DOWN_ACTION) else None
+        self._volume_set_action = Script(hass, config.get(
+            CONF_VOLUME_SET_ACTION)) if config.get(
+            CONF_VOLUME_SET_ACTION) else None
+        self._volume_mute_action = Script(hass, config.get(
+            CONF_VOLUME_MUTE_ACTION)) if config.get(
+            CONF_VOLUME_MUTE_ACTION) else None
+        self._volume_unmute_action = Script(hass, config.get(
+            CONF_VOLUME_UNMUTE_ACTION)) if config.get(
+            CONF_VOLUME_UNMUTE_ACTION) else None
+        self._select_source_action = Script(hass, config.get(
+            CONF_SELECT_SOURCE_ACTION)) if config.get(
+            CONF_SELECT_SOURCE_ACTION) else None
+        self._sources = config.get(CONF_SOURCE_NAMES)
+        self._selected_source = None
+        self._state = STATE_OFF
+        self._volume = config.get(CONF_INITIAL_VOLUME)
+        self._volume_step = config.get(CONF_VOLUME_STEP)
+        self._volume_step_delay = config.get(CONF_VOLUME_STEP_DELAY)
+        self._is_muted = False
+
+    @property
+    def name(self):
+        """Return the name of the device."""
+        return self._name
+
+    @property
+    def supported_features(self):
+        """Flag media player features that are supported."""
+        features = 0
+        if self._turn_on_action:
+            features |= SUPPORT_TURN_ON
+        if self._turn_off_action:
+            features |= SUPPORT_TURN_OFF
+        if (self._volume_up_action and self._volume_down_action) or \
+                self._volume_set_action:
+            features |= SUPPORT_VOLUME_SET
+        if self._volume_mute_action and self._volume_unmute_action:
+            features |= SUPPORT_VOLUME_MUTE
+        if self._sources and self._select_source_action:
+            features |= SUPPORT_SELECT_SOURCE
+        return features
+
+    @property
+    def state(self):
+        """Return the state of the device."""
+        return self._state
+
+    @property
+    def source_list(self):
+        """List of available input sources."""
+        return list(self._sources.values())
+
+    def turn_on(self):
+        """Turn on media player."""
+        self._state = STATE_ON
+        if self._turn_on_action:
+            self._turn_on_action.run()
+
+    def turn_off(self):
+        """Turn off media player."""
+        self._state = STATE_OFF
+        if self._turn_off_action:
+            self._turn_off_action.run()
+
+    @property
+    def volume_level(self):
+        """Volume level of the media player (0..1)."""
+        return self._volume / 100.0
+
+    def set_volume_level(self, volume):
+        volume = volume * 100
+        volume = int(math.ceil(volume / 10.0)) * 10
+        if self._volume_set_action:
+            self._volume = volume
+            self._volume_set_action.run({'volume:': volume})
+        elif self._volume_up_action and self._volume_down_action:
+            if volume < self._volume:
+                while volume < self._volume:
```

Comment:
We should not fake things, we should only execute script/services and they can do it if the user want it

Reply:
Given this use cases for this component, I think this is a good feature. But, maybe the volume stepping should be configurable turn this behaviour on or off. Eg:
```yaml
media_player:
  - platform: template_media_player
    name: "The Amp"
    use_volume_up_and_down_to_set_volume: true
    volume_up_action:
      service: logbook.log
      data:
        name: Delegate media media_player
        message: Crank it up!
    volume_down_action:
      service: logbook.log
      data:
        name: Delegate media media_player
        message: Turn it down!
```